### PR TITLE
#4358: Use standard memoize instead of custom storage logic

### PR DESCRIPTION
**Pull request by fregante**
## What does this PR do?

- part of https://github.com/pixiebrix/pixiebrix-extension/issues/4358

I'm looking for and replacing all instances of data that should be stored in `chrome.storage.session`. This brought me to replace a couple of patterns around `getToken` and `ensureContextMenus`

## Checklist

- [x] Add tests
- [ ] Run Storybook and manually confirm that all stories are working
- [x] Designate a primary reviewer: @twschiller

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -652,3 +639,23 @@ export function isValidUrl(
...
+ * @example
+ *   const memFetch = ignoreRepeatedCalls(fetch)
+ *   await Promise([memFetch('/time'), memFetch('/time')])
+ *   // => both will return the exact same Promise
+ *   await memFetch('/time')
+ *   // => no concurrent calls at this time, so another request made
+ *
+ * @see https://github.com/sindresorhus/promise-fun/issues/15
+ */
+export const memoizeUntilSettled: typeof pMemoize = (
```
Comment (twschiller) [inline src/utils.ts:654]: 👍

File: src/utils.ts [hunk 2]
```diff
@@ -265,16 +262,6 @@ export function ensureJsonObject(value: Record<string, unknown>): JsonObject {
   return JSON.parse(safeJsonStringify(value)) as JsonObject;
 }
 
-export function clearObject(obj: Record<string, unknown>): void {
```
Comment (twschiller) [inline src/utils.ts:268]: 👍

File: src/utils.ts [hunk 3]
```diff
@@ -249,10 +250,6 @@ export function boolean(value: unknown): boolean {
   return false;
 }
 
-export function clone<T extends Record<string, unknown>>(object: T): T {
```
Comment (twschiller) [inline src/utils.ts:252]: 👍

File: src/background/auth.ts [hunk 4]
```diff
@@ -98,18 +99,26 @@ export async function deleteCachedAuthData(serviceAuthId: UUID): Promise<void> {
...
 /**
- * @see getToken
+ * Exchange credentials for a token, and cache the token response.
+ *
+ * If a request for the token is already in progress, return the existing promise.
  */
-export async function _getToken(
+
+export const getToken = memoizeUntilSettled(_getToken, {
+  cacheKey: ([, auth]) => auth.id,
```
Comment (twschiller) [inline src/background/auth.ts:109]: 👍

File: src/background/contextMenus.ts [hunk 5]
```diff
@@ -154,7 +137,10 @@ export async function uninstallContextMenu({
   }
 }
 
-export async function ensureContextMenu({
+export const ensureContextMenu = memoizeUntilSettled(_ensureContextMenu, {
```
Comment (twschiller) [inline src/background/contextMenus.ts:140]: 👍

File: src/background/contextMenus.ts [hunk 6]
```diff
@@ -132,16 +121,7 @@ export async function uninstallContextMenu({
...
-    const menuItemId = extensionMenuItems.get(extensionId);
-
-    if (menuItemId) {
-      extensionMenuItems.delete(extensionId);
-      await browser.contextMenus.remove(menuItemId);
-    } else {
-      // Our bookkeeping in `extensionMenuItems` might be off. Try removing the expected menuId just in case.
-      await browser.contextMenus.remove(makeMenuId(extensionId));
-    }
-
```
Comment (fregante) [inline src/background/contextMenus.ts:144]: There are two changes in this file:

- avoid storing the result of `makeMenuId(extensionId)` and just create it as needed
- simplify logic around create/update, until this makes it easier:
	- https://github.com/w3c/webextensions/issues/312

**Review by fregante — COMMENTED** [review]
I summarily tested multiple context menus:
- opened multiple tabs
- opened and closed the dev tools
- refreshed the page
- deleted some context menus (and checked in another still-open tab)
- …

No errors anywhere, the right amount of context menus exist.

Screen Shot

**Review by twschiller — APPROVED** [review]

**Pull request merged**